### PR TITLE
Request for review, merge for Muf3.0 rc1 changes

### PR DESCRIPTION
**Pull request by sushma-ananda**
HI @timobaumann , as per your review comments from today's meeting, I have updated #commit b38bc0f3f748307f8f31aeac51c3e9054ac6580b .
can you please provide your review comments for https://github.com/fpecune/MUF-inMind-April2017-master/commits/muf3.0RC1Changes ?

### CODE REVIEW

**Comment by sushma-ananda** [comment]
@timobaumann I uploaded a fix : ba38960a9d6fc6e4100b747d3c1251dd87f71027 and with this fix I am able to run an iteration from greeting till recommend phase.

File: .gitignore [hunk 1]
```diff
@@ -7,10 +7,15 @@ stderr.*
 .DS_Store
 /build
 /saracommons/build
+saracommons/out/
 /saracsc/build
 /sarabeat/build
 /captures
 /.gradle
+saracommons/out
+saracommons/out/*
```
Comment (timobaumann) [inline .gitignore]: do we need both saracommons/out *and* saracommons/out/*?
likewise for build below: also, isn't that already covered by build  above?

File: XMLData [hunk 2]
```diff
@@ -1 +0,0 @@
-sarabeat/XMLData/
```
Comment (timobaumann) [inline XMLData]: could you turn this (and sarabeat/Database) into links again? I don't understand the reason for taking them out.

File: build.gradle [hunk 3]
```diff
@@ -52,11 +52,14 @@ configurations.all {
...
     testCompile "org.mockito:mockito-core:1.+"
-    compile 'edu.cmu.inmind.framework:multiuser:2.15'
+    compile 'edu.cmu.inmind.framework.multiuser:server:3.0-rc1:all'
     // version
     compile project(":saracommons")
     compile project(":saracsc")
     compile project(":sarabeat")
+    //compile project(":MultiUserFramework")
+    //compile group: 'edu.cmu.inmind.framework.multiuser:server:3.0-rc1:all'
+    //compile fileTree(dir: 'libs', include: ['muf-all-3.0-rc1-all.jar'])
```
Comment (timobaumann) [inline build.gradle]: I think these lines are superseeded by the MUF include at the top, right? → remove.

File: services.json [hunk 4]
```diff
@@ -12,8 +12,8 @@
     {
       "serviceName": "NLU_DM_SERVICE",
       "__comment" : "set this address to where your python-dialogue manager lives.",
-      "serverAddress": "tcp://34.203.160.208:5590",
-      "clientAddress": "tcp://34.203.160.208:5590",
+      "masterMUFAddress": "tcp://34.203.160.208:5660",
+      "slaveMUFAddress": "tcp://34.203.160.208:5660",
```
Comment (timobaumann) [inline services.json]: please set to old address

File: src/main/java/edu/cmu/inmind/multiuser/sara/MainClassBase.java [hunk 5]
```diff
@@ -48,9 +48,11 @@ public void run() {
         while (!command.equals("shutdown")) {
             System.err.println("Type \"shutdown\" to stop:");
```
Comment (timobaumann) [inline src/main/java/edu/cmu/inmind/multiuser/sara/MainClassBase.java]: while you're at it, maybe change to 
System.err.println("Type " + SaraCons.SHUTDOWN + " to stop:");

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/CSCComponent.java [hunk 6]
```diff
@@ -79,11 +82,11 @@ public void onEvent(BlackboardEvent blackboardEvent) throws Exception {
         List<Strategy> strategies = new ArrayList<Strategy>();
 
         for (ConversationalStrategy cs : EnumSet.of(SD, SE, Praise, QESD, VSN, ASN)) {
-            strategies.add(new Strategy(cs.shortName(), strategyList.get(cs)));
+            strategies.add(new Strategy(cs.name(), strategyList.get(cs)));
```
Comment (timobaumann) [inline src/main/java/edu/cmu/inmind/multiuser/sara/component/CSCComponent.java]: I think this needs to remain shortName, because there's extra stuff necessary for Praise (which in other parts is called PR).

File: services.json (hunk 4 above)
Comment (sushma-ananda) [reply services.json]: fixed in commit # b38bc0f3f748307f8f31aeac51c3e9054ac6580b

File: XMLData (hunk 2 above)
Comment (sushma-ananda) [reply XMLData]: I think they were on the latest commit muf3.0RC1Changes as I tested now. They should work now. Please let me know if latest version does not work for this one.

File: .gitignore (hunk 1 above)
Comment (sushma-ananda) [reply .gitignore]: fixed in commit # 821ce35

File: build.gradle (hunk 3 above)
Comment (sushma-ananda) [reply build.gradle]: fixed in commit # 821ce35

File: src/main/java/edu/cmu/inmind/multiuser/sara/MainClassBase.java (hunk 5 above)
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/MainClassBase.java]: fixed in commit #821ce35

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/CSCComponent.java (hunk 6 above)
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/CSCComponent.java]: fixed in commit #821ce35

**Comment by sushma-ananda** [comment]
@timobaumann Hi Timo, did you happened get a chance to review the changes to merge ?

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java [hunk 7]
```diff
@@ -24,7 +25,7 @@
 public class NLU_DMComponent extends PluggableComponent {
     private ActiveDMOutput dmOutput;
 
-    private Blackboard blackboard;
+    protected static Blackboard blackboard;
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java]: my bad. I have added it for my tests for testing something a while ago for fixing some other bug. But once I complete my tests , I will upload to fpecune/MUF-inMind-April2017 repo back shortly. You can make it normal private variable, which works as well. Anyhow, let me know if you were going to need these changes asap. Thanks.

[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java]: my bad. I have added it for my tests for testing something a while ago for fixing some other bug. But once I complete my tests , I will upload to fpecune/MUF-inMind-April2017 repo back shortly. You can make it normal private variable, which works as well. And yes we do need blackboard, as per last 3-4 meetinsg we had on MUF, there is no more blackboard() and every event is going to send in blackboard instance. So, since you are using blackboard extensively in NLU_DMComponent, it is better that Blackboard variable be there. It does not matter if it is static or not. Anyhow, please let me know if you were going to need these changes asap. Thanks.

File: src/main/java/edu/cmu/inmind/multiuser/sara/MainClassBase.java [hunk 8]
```diff
@@ -48,9 +48,11 @@ public void run() {
...
             Scanner scanner = new Scanner(System.in);
-            command = scanner.nextLine();
-            if (command.equals(SaraCons.SHUTDOWN)) {
-                MUFLifetimeManager.stopFramework( muf );
+            if(scanner.hasNextLine()) {
```
Comment (timobaumann) [inline src/main/java/edu/cmu/inmind/multiuser/sara/MainClassBase.java]: with this change, the outer while-loop indefinitely  prints "Type shutdown to stop:". What was the reason for this change, was there anything wrong before?

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java]: @timobaumann it seems someone implemented ActiveDMOutput that is calling static method sendToBBforSession . So it seem this forces blackboard object to be static. all for the single call in processQueryResponse() :         **dmOutput.plainGetRecommendation().setRexplanations(Utils.fromJson(message, DMOutput.class).getRecommendation().getRexplanations());**
But it is not the best practice. do you know why they want to use static sendBB?

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java [hunk 9]
```diff
@@ -34,7 +35,12 @@ static void sendToBBforSession(String sessionID, String msgId, ActiveDMOutput dm
         NLU_DMComponent nluc = components.get(sessionID);
         nluc.dmOutput = dmoutput;
         Log4J.debug(nluc, "the local activedmoutput is " + nluc.dmOutput.hashCode());
-        nluc.blackboard.post(nluc, msgId, "query");
+        try {
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java]: fixed in commit: 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/FakeCSCComponent.java [hunk 10]
```diff
@@ -39,23 +40,23 @@ public void postCreate(){
...
+    public void onEvent(Blackboard blackboard,BlackboardEvent blackboardEvent) throws Throwable
     {
         CSCOutput cscOutput = new CSCOutput();
         Random r = new Random();
 
         List<Strategy> strategyList = new ArrayList<Strategy>();
 
         for (ConversationalStrategy cs : EnumSet.of(SD, SE, Praise, QESD, VSN, ASN)) {
-            strategyList.add(new Strategy(cs.shortName(), r.nextDouble()));
+            strategyList.add(new Strategy(cs.name(), r.nextDouble()));
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/FakeCSCComponent.java]: fixed in 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLGComponent.java [hunk 11]
```diff
@@ -98,7 +111,13 @@ public void receiveMessage(String msg) {
          * update the blackboard
          */
         Log4J.info(this, "SessionID: " + this.getSessionId() + "Output: " +bson.getSpeech());
-        blackboard().post( this, SaraCons.MSG_NLG, bson );
+        try {
+            getBlackBoard(getSessionId()).post(this, SaraCons.MSG_NLG, bson);
+        }catch (Throwable t)
+        {
+            t.printStackTrace();
+
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLGComponent.java]: 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/FakeCSCComponent.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/FakeCSCComponent.java]: fixed in 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLGComponent.java [hunk 12]
```diff
@@ -52,7 +54,13 @@ public void execute() {
 
     @Loggable
     private void extractAndProcess() {
-        SROutput srOutput = (SROutput) blackboard().get(SaraCons.MSG_SR);
+        SROutput srOutput = null;
+        try {
+            srOutput = (SROutput) blackboardEvent.getElement();
+        }catch (Throwable t)
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLGComponent.java]: Blackboard.post and underlying methods itself throws exception since we have testing on rc1 for now. so, it seems it is better we do have the stacktrace. Please do check: 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: services.json (hunk 4 above)
Comment (timobaumann) [reply services.json]: thanks.

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLGComponent.java [hunk 13]
```diff
@@ -78,8 +86,13 @@ private void extractAndProcess() {
      * processes in parallel rather than sequentially.
      */
     @Override
-    public void onEvent(BlackboardEvent event) throws Exception
+    public void onEvent(Blackboard blackboard, BlackboardEvent event) throws Exception
     {
+        if(blackboard!=null) {
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLGComponent.java]: There is no more any reason from our tests/fixes for blackboard to be null: we are optimistic and also 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLGComponent.java [hunk 14]
```diff
@@ -98,7 +111,13 @@ public void receiveMessage(String msg) {
          * update the blackboard
          */
         Log4J.info(this, "SessionID: " + this.getSessionId() + "Output: " +bson.getSpeech());
-        blackboard().post( this, SaraCons.MSG_NLG, bson );
+        try {
```
[the remark this answers is not in the document]
Comment (timobaumann) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLGComponent.java]: alright. can you make a note/bug for MUF that blackboard.post() should not throw a throwable (which is the least specific thing that could be thrown) but something more specific?

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLGComponent.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLGComponent.java]: 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java [hunk 15]
```diff
@@ -84,25 +110,36 @@ public void onEvent(BlackboardEvent blackboardEvent) throws Exception {
                     "DM intent process thread").start();
         } else if (blackboardEvent.getId().equals(SaraCons.MSG_ASR)) {
             Log4J.debug(this, "sending on " + blackboardEvent.toString());
-            blackboard().post(this, SaraCons.MSG_ASR_DM, blackboardEvent.getElement());
+            NLU_DMComponent.blackboard.post(this, SaraCons.MSG_ASR_DM, blackboardEvent.getElement());
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java]: implemented in 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java [hunk 16]
```diff
@@ -84,25 +110,36 @@ public void onEvent(BlackboardEvent blackboardEvent) throws Exception {
...
+
         } else {
             Log4J.error(this, "got a message I could not understand: " + blackboardEvent.toString());
         }
     }
 
     private void processStartDM() {
         Log4J.debug(this, "about to send initial greeting ...");
-        blackboard().post(this, SaraCons.MSG_START_DM_PYTHON, "");
+        try {
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java]: implemented in 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java [hunk 17]
```diff
@@ -84,25 +110,36 @@ public void onEvent(BlackboardEvent blackboardEvent) throws Exception {
...
+        }
         Log4J.debug(this, "Sent Initial Greeting");
     }
 
     private void processUserModel(Object element) {
         final UserModel userModel = (UserModel) element;
         Log4J.info(this, "Received user model");
         if (userModel.getUserFrame() != null) {
-            Log4J.info(this, "Sending user frame to python: " + Utils.toJson(userModel.getUserFrame()));
+            Log4J.info(this, "Sending user frame to python: " + edu.cmu.inmind.multiuser.controller.common.Utils.toJson(userModel.getUserFrame()));
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java]: implemented in 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java [hunk 18]
```diff
@@ -84,25 +110,36 @@ public void onEvent(BlackboardEvent blackboardEvent) throws Exception {
...
             // no reply for user model stuff
-            blackboard().post(this, SaraCons.MSG_USER_FRAME, userModel.getUserFrame());
+            try {
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java]: implemented in 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java [hunk 19]
```diff
@@ -111,13 +148,18 @@ private void processUserModel(Object element) {
     private void processDMIntent(String message) {
         Log4J.debug(NLU_DMComponent.this, "I've received python response: " + message);
         // store user's utterance (for NLG)
-        dmOutput = Utils.fromJson(message, ActiveDMOutput.class);
+        dmOutput = edu.cmu.inmind.multiuser.controller.common.Utils.fromJson(message, ActiveDMOutput.class);
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java]: implemented in 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java [hunk 20]
```diff
@@ -111,13 +148,18 @@ private void processUserModel(Object element) {
...
         /* uncomment the next two lines for incrementality: */
         if (dmOutput.plainGetRecommendation() != null)
             dmOutput.plainGetRecommendation().setRexplanations(null);
         dmOutput.sessionID = getSessionId();
         // post to Blackboard
-        blackboard().post(NLU_DMComponent.this, SaraCons.MSG_DM, dmOutput);
+        try {
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLU_DMComponent.java]: fixed in 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/SocialReasonerComponent.java [hunk 21]
```diff
@@ -38,9 +39,14 @@ public void execute() {
     }
 
 
-    private void updateRapport(){
-        RapportOutput rapportOutput = edu.cmu.inmind.multiuser.common.Utils.fromJson(
-                (String) blackboard().get(SaraCons.MSG_RPT), RapportOutput.class );
+    private void updateRapport(BlackboardEvent blackboardEvent){
+        RapportOutput rapportOutput = null;
+        try {
+            rapportOutput = Utils.fromJson(
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/SocialReasonerComponent.java]: implemented in 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/SocialReasonerComponent.java [hunk 22]
```diff
@@ -49,16 +55,26 @@ private void updateRapport(){
         Log4J.info(this,"RapporEstimator : " + rapport );
     }
 
-    private void updateStrategy(){
-        CSCOutput csc = (CSCOutput) blackboard().get(SaraCons.MSG_CSC);
+    private void updateStrategy(BlackboardEvent event){
+        CSCOutput csc = null;
+        try {
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/SocialReasonerComponent.java]: implemented in 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/SocialReasonerComponent.java [hunk 23]
```diff
@@ -108,27 +131,32 @@ private SROutput selectStrategy(){
...
+    public void onEvent(Blackboard blackboard, BlackboardEvent event) throws Throwable {
         //TODO: add code here
         //...
         Log4J.info(this, "SocialReasonerComponent. These objects have been updated at the blackboard: "
                 + event.toString());
-
+        if(blackboard==null)
+        {
+            Log4J.info(this, "blackboard is null");
+        }
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/SocialReasonerComponent.java]: There is more any reason for blackboard to be null. As mentioned to you during earlier review comments, this was added for testing. also 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/MainClassBase.java (hunk 8 above)
Comment (timobaumann) [reply src/main/java/edu/cmu/inmind/multiuser/sara/MainClassBase.java]: well, for me, what happens is the following: I haven't typed anything, hence scanner.hasNextLine() returns false and the while-loop keeps churning, printing the line over and over again. Feel free to drop by to see it fill the screen in no time. 

I don't see how the screenshot adds anything.

It's not helpful to argue about fixing one bug which then exposes a different bug.

The underlying problem may be different: code reviews are nothing personal and not about assigning blame. I appreciate your changes and it's not about who wrote the while loop. It's about the fact that you're in that area anyway so fixing it for good is what needs to be done.

why don't you try 
```java
        while (!command.equals(SaraCons.SHUTDOWN)) {
            if(scanner.hasNextLine()) {
                command = scanner.nextLine();
                if (command.equals(SaraCons.SHUTDOWN)) {
                    MUFLifetimeManager.stopFramework(muf);
                }
                System.err.println("Type " + SaraCons.SHUTDOWN + " to stop:");
            } else {
                Utils.sleep(300);
            }
        }
```
instead of arguing about who is right or wrong?

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/SocialReasonerComponent.java [hunk 24]
```diff
@@ -108,27 +131,32 @@ private SROutput selectStrategy(){
...
         if (event.getId().equals(SaraCons.MSG_UM)) {
             updateUserModel(((UserModel) event.getElement()));
         }
         if (event.getId().equals(SaraCons.MSG_DM)) {
-            sendToNLG = selectStrategy();
-            blackboard().post(this, SaraCons.MSG_SR, sendToNLG);
+            if(blackboard!=null)
+                sendToNLG = selectStrategy( event);
+
+            blackboard.post(this, SaraCons.MSG_SR, sendToNLG);
```
[the remark this answers is not in the document]
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/SocialReasonerComponent.java]: There is more any reason for blackboard to be null. As mentioned to you during earlier review comments, this was added for testing. also 4cbfc58323e2ad7b0b42f31af2f73fc50df213a9

File: src/main/java/edu/cmu/inmind/multiuser/sara/MainClassBase.java (hunk 8 above)
Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/MainClassBase.java]: I appreciate your comments. Thanks for your carefully reviewed comments. As much I would like to discuss more and fix this issue for you guys, I also believe do not judge my skills based on other's written code. Thanks.

Comment (sushma-ananda) [reply src/main/java/edu/cmu/inmind/multiuser/sara/MainClassBase.java]: @timobaumann I tried your solution. I still dont get more than one print statement of "Shutdown". Please, when you get a chance, share details of exact scenario when this happens ? Also it will be good to know if anyone else getting same issue.

File: src/main/java/edu/cmu/inmind/multiuser/sara/component/NLGComponent.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (ojrlopez27) [reply src/main/java/edu/cmu/inmind/multiuser/sara/component/NLGComponent.java]: Done.
Note: ZMQ library throws both Exceptions and Errors. Most of them are caught and printed out in the console by ZMQ, so we don't have control to do alternative flows when those exceptions and errors happen. We have modified ZMQ code so those exceptions and errors are thrown to the highest layers instead of printing out in console.
